### ISSUE TITLE
Allow linters to provide fix

### ISSUE THREAD

**Issue by renkun-ken** [jimhester/lintr#784]
Some code editors (e.g. vscode) supports [code actions](https://code.visualstudio.com/docs/editor/refactoring) to allow language server to provide code refactoring or fixes via the [language server protocol](https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_codeAction).

In languageserver, I'm trying to support code actions like this to provide quick fixes of the lint results for many linters. However, lintr might be the best place to provide the fix along with the problem as each linter has the best knowledge of the problem it finds and how it should be resolved if possible.

Does it make sense to allow an optional a field `fix` in the lint result which contains a list of edits (each has a range and a new text) so that languageserver could directly transform these fixes into code actions at downstream?

The code issues found by many linters could be easy to fix and the fix is optional rather than required so that implementing fixes could be done in a progressive way.

Also see https://github.com/REditorSupport/languageserver/issues/396.

**Comment by AshesITR** [comment]
This sounds promising. Maybe we could upgrade the `ranges` attribute somehow?
IINM it's possible to specify multiple ranges for a single lint and they already contain start and end markers.
The current interface assumes a list of length-2 integer vectors.

We could add optional attributes to each vector (e.g. attr(., "fix")) or add a new attribute to `Lint()`.
Using the names() of ranges would also be thinkable, but that will lead to issues when the source code is in a non-system encoding as R names must be encoded in the system encoding.

**Comment by renkun-ken** [comment]
A simple way is to add `fixes` to `Lint` structure like the following:

```r
  structure(
    list(
      filename = filename,
      line_number = as.integer(line_number),
      column_number = as.integer(column_number),
      type = type,
      message = message,
      line = line,
      ranges = ranges,
      fixes = fixes,
      linter = NA_character_
    ),
    class = "lint")
```

where `fixes` is a character vector of replacement strings for each range in `ranges`. This makes it the ranges to emphasize must be the ranges to replace text as fixes.

**Comment by renkun-ken** [comment]
Here's a simple demo with #785 and https://github.com/REditorSupport/languageserver/pull/397:


![Kapture 2021-03-21 at 23 35 50](https://user-images.githubusercontent.com/4662568/111911030-b5883300-8a9e-11eb-85b5-a5ff2b03953d.gif)

**Comment by MichaelChirico** [comment]
FWIW this is exactly the sort of thing I had in mind for #710
